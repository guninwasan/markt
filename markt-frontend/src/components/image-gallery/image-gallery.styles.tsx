import styled from "styled-components";
import { colors } from "../../utils";

const GalleryContainer = styled.div`
  display: flex;
  gap: 1rem;
  justify-content: center;
  margin-top: 10px;
  flex-wrap: wrap;
  user-select: none;
  user-drag: none;
`;

const ModalContainer = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.8);
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 1000;
`;

const CarouselContent = styled.div`
  position: relative;
  max-width: 70%;
  max-height: 80vh;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const ImageWrapper = styled.div`
  position: relative;
  width: 200px;
  height: 200px;
  object-fit: contain;
  overflow: hidden;
  cursor: pointer;
  border-radius: 8px;
  transition: transform 0.3s ease;
  &:hover {
    transform: scale(1.05);
  }
  user-select: none;
  user-drag: none;
`;

const MediaWrapper = styled.div`
  width: 100%;
  height: 100%;
  max-height: 70vh;
  object-fit: contain;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const ZoomedImage = styled.img<{
  offsetX?: number;
  offsetY?: number;
}>`
  width: 100%;
  height: 70vh;
  object-fit: contain;
  transition: transform 0.3s ease;
  user-select: none;
  user-drag: none;
`;

const ThumbnailContainer = styled.div`
  display: flex;
  gap: 10px;
  margin-top: 20px;
  justify-content: center;
  user-select: none;
  user-drag: none;
`;

const Thumbnail = styled.div<{ active: boolean }>`
  width: 60px;
  height: 60px;
  overflow: hidden;
  border: ${({ active }) =>
    active ? `2px solid ${colors.white}` : "2px solid transparent"};
  transition: transform 0.2s ease;
  cursor: pointer;
  &:hover {
    transform: scale(1.05);
  }
  user-select: none;
  user-drag: none;
`;

const ThumbnailImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  user-select: none;
  user-drag: none;
`;

const ThumbnailVideo = styled.video`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const NavButton = styled.button`
  background: rgba(255, 255, 255, 0.6);
  border: none;
  font-size: 30px;
  color: ${colors.darkGrey};
  cursor: pointer;
  padding: 10px;
  border-radius: 50%;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  transition: background 0.3s ease;
  &:hover {
    cursor: pointer;
    background: rgba(255, 255, 255, 0.9);
  }
`;

const PrevButton = styled(NavButton)`
  left: -100px;
`;

const NextButton = styled(NavButton)`
  right: -100px;
`;

const CloseButton = styled.button`
  position: absolute;
  top: 20px;
  right: 20px;
  background: none;
  border: none;
  color: ${colors.white};
  font-size: 30px;
  cursor: pointer;
  &:hover {
    color: ${colors.red};
  }
`;

const IndexText = styled.p`
  color: ${colors.white};
  font-size: 12px;
`;

export {
  GalleryContainer,
  ModalContainer,
  CarouselContent,
  ImageWrapper,
  MediaWrapper,
  ZoomedImage,
  ThumbnailContainer,
  Thumbnail,
  ThumbnailImage,
  ThumbnailVideo,
  NavButton,
  PrevButton,
  NextButton,
  CloseButton,
  IndexText,
};
