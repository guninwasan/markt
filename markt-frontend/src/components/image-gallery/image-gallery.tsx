import React, { useState } from "react";
import {
  GalleryContainer,
  ImageWrapper,
  ThumbnailImage,
  ThumbnailVideo,
  ModalContainer,
  CloseButton,
  CarouselContent,
  PrevButton,
  NextButton,
  MediaWrapper,
  ThumbnailContainer,
  Thumbnail,
  ZoomedImage,
  IndexText,
} from "./image-gallery.styles";

const detectMediaType = (url: string): "image" | "video" => {
  const imageExtensions = [".jpg", ".jpeg", ".png", ".gif"];
  const videoExtensions = [".mp4", ".webm", ".ogg"];
  const extension = url.split(".").pop()?.toLowerCase();
  return extension && videoExtensions.includes(`.${extension}`)
    ? "video"
    : "image";
};

interface ImageGalleryProps {
  mediaUrls: string[];
}

const ImageGallery: React.FC<ImageGalleryProps> = ({ mediaUrls }) => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [currentIndex, setCurrentIndex] = useState(0);
  const [zoomPosition, setZoomPosition] = useState({ x: 0, y: 0 });

  const mediaItems = mediaUrls.map((url) => ({
    src: url,
    type: detectMediaType(url),
  }));

  const handleMediaClick = (index: number) => {
    setCurrentIndex(index);
    setIsModalOpen(true);
    setZoomPosition({ x: 0, y: 0 });
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  const goToNextMedia = (e: React.MouseEvent) => {
    e.stopPropagation();
    setCurrentIndex((prev) => (prev + 1) % mediaUrls.length);
    setZoomPosition({ x: 0, y: 0 });
  };

  const goToPrevMedia = (e: React.MouseEvent) => {
    e.stopPropagation();
    setCurrentIndex((prev) => (prev === 0 ? mediaUrls.length - 1 : prev - 1));
    setZoomPosition({ x: 0, y: 0 });
  };

  // const handleMouseMove = (e: any) => {
  //   const { offsetX, offsetY, target } = e.nativeEvent;
  //   const { width, height } = target;
  //   const offsetXPercent = ((offsetX / width) * 2 - 1) * 25;
  //   const offsetYPercent = ((offsetY / height) * 2 - 1) * 25;
  //   setZoomPosition({ x: offsetXPercent, y: offsetYPercent });
  // };

  return (
    <>
      <GalleryContainer>
        {mediaItems.slice(0, 4).map((item, index) => (
          <ImageWrapper key={index} onClick={() => handleMediaClick(index)}>
            {item.type === "image" ? (
              <ThumbnailImage src={item.src} />
            ) : (
              <ThumbnailVideo src={item.src} controls muted />
            )}
          </ImageWrapper>
        ))}
      </GalleryContainer>

      {isModalOpen && (
        <ModalContainer>
          <CloseButton onClick={closeModal}>{"✕"}</CloseButton>
          <CarouselContent>
            <PrevButton onClick={goToPrevMedia} data-testid="prev">
              {"❮"}
            </PrevButton>
            {/* <MediaWrapper onMouseMove={handleMouseMove}> */}
            <MediaWrapper>
              {mediaItems[currentIndex].type === "image" ? (
                <ZoomedImage
                  src={mediaItems[currentIndex].src}
                  offsetX={zoomPosition.x}
                  offsetY={zoomPosition.y}
                />
              ) : (
                <ThumbnailVideo
                  src={mediaItems[currentIndex].src}
                  controls
                  autoPlay
                />
              )}
            </MediaWrapper>
            <NextButton onClick={goToNextMedia} data-testid="next">
              {"❯"}
            </NextButton>
            <ThumbnailContainer>
              {mediaItems.map((item, index) => (
                <Thumbnail
                  key={index}
                  active={index === currentIndex}
                  onClick={() => setCurrentIndex(index)}
                >
                  {item.type === "image" ? (
                    <ThumbnailImage src={item.src} />
                  ) : (
                    <ThumbnailVideo src={item.src} muted />
                  )}
                </Thumbnail>
              ))}
            </ThumbnailContainer>
            <IndexText>
              {currentIndex + 1}/{mediaItems.length}
            </IndexText>
          </CarouselContent>
        </ModalContainer>
      )}
    </>
  );
};

export { ImageGallery };
