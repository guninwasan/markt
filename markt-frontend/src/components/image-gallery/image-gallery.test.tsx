import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { ImageGallery } from "./image-gallery";

const mediaUrls = [
  "https://example.com/image1.jpg",
  "https://example.com/image2.jpg",
];

describe("ImageGallery", () => {
  it("should render thumbnails", () => {
    render(<ImageGallery mediaUrls={mediaUrls} />);
    expect(screen.getAllByRole("img")).toHaveLength(2);
  });

  it("should open modal on thumbnail click", () => {
    render(<ImageGallery mediaUrls={mediaUrls} />);

    fireEvent.click(screen.getAllByRole("img")[0]);
    expect(screen.getByText("✕")).toBeInTheDocument();
  });

  it("should close modal on close button click", () => {
    render(<ImageGallery mediaUrls={mediaUrls} />);

    fireEvent.click(screen.getAllByRole("img")[0]);
    fireEvent.click(screen.getByText("✕"));
    expect(screen.queryByText("✕")).not.toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { container } = render(<ImageGallery mediaUrls={mediaUrls} />);
    expect(container).toMatchSnapshot();
  });

  it("should navigate to next image", () => {
    render(<ImageGallery mediaUrls={mediaUrls} />);
    fireEvent.click(screen.getAllByRole("img")[0]);

    const nextButton = screen.getByTestId("next");
    expect(nextButton).toBeInTheDocument();

    const getFirstImageIndex = screen.getByText("1/2");
    expect(getFirstImageIndex).toBeInTheDocument();
    fireEvent.click(nextButton);

    const getSecondImageIndex = screen.getByText("2/2");
    expect(getSecondImageIndex).toBeInTheDocument();
  });

  it("should navigate to previous image", () => {
    render(<ImageGallery mediaUrls={mediaUrls} />);
    fireEvent.click(screen.getAllByRole("img")[1]);

    const prevButton = screen.getByTestId("prev");
    expect(prevButton).toBeInTheDocument();

    const getSecondImageIndex = screen.getByText("2/2");
    expect(getSecondImageIndex).toBeInTheDocument();
    fireEvent.click(prevButton);

    const getFirstImageIndex = screen.getByText("1/2");
    expect(getFirstImageIndex).toBeInTheDocument();
  });
});
