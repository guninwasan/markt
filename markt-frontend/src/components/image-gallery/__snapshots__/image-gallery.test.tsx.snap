// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery should match snapshot 1`] = `
<div>
  <div
    class="sc-blHHSb kscHbD"
  >
    <div
      class="sc-fAUdSK jIA-DAh"
    >
      <img
        class="sc-ixGGxD fzPTpc"
        src="https://example.com/image1.jpg"
      />
    </div>
    <div
      class="sc-fAUdSK jIA-DAh"
    >
      <img
        class="sc-ixGGxD fzPTpc"
        src="https://example.com/image2.jpg"
      />
    </div>
  </div>
</div>
`;
